![data-x](http://oi64.tinypic.com/o858n4.jpg)

---
# Pandas Introduction 
### Pandas BKHW with Stock Data and Correlation Examples


**Author list:** Ikhlaq Sidhu

**References / Sources:** 
Includes examples from Wes McKinney


**License Agreement:** Feel free to do whatever you want with this code

___

## What Does Pandas Do?
<img src="https://github.com/ikhlaqsidhu/data-x/raw/master/imgsource/pandas-p1.jpg">


## What is a Pandas Table Object?
<img src="https://github.com/ikhlaqsidhu/data-x/raw/master/imgsource/pandas-p2.jpg">


## This table is a dictionary of sequences (like np arrays)
<img src="https://github.com/ikhlaqsidhu/data-x/raw/master/imgsource/pandas-p3.jpg">



### Creating a Data Frame
**Key Points:** Pandas has Series (like Arrays), DataFrames (like Tables), and Panels (3D version)

In [1]:
import pandas as pd
import numpy as np
#pd.show_versions()
pd.__version__

'0.20.3'

### Our first goal: Learn that it is easy to create a data frame

#### We use pandas.DataFrame and put in just about anything data type as an argument

class pandas.DataFrame(data=None, index=None, columns=None, dtype=None, copy=False)

data ca be a numpy ndarray (structured or homogeneous), dict, or DataFrame. 
Dict can contain Series, arrays, constants, or list-like objects

In [2]:
# Try it with an array
a1 = np.array(np.random.randn(3))
a2 = np.array(np.random.randn(3))
a3 = np.array(np.random.randn(3))

print (a1, type(a1))
print (a2)
print (a3)

[ 0.15660932 -1.55311436 -1.15361166] <class 'numpy.ndarray'>
[ 0.93958342 -0.68420685 -0.47135349]
[-1.70427419 -1.91649166  1.18210322]


In [3]:
df0 = pd.DataFrame(a1)
print ("This is a dataframe")
df0

This is a dataframe


0
0  0.156609
1 -1.553114
2 -1.153612

In [5]:
# What if I give it a list of np arrays?
df0 = pd.DataFrame([a1, a2, a3])
print (type(df0))
df0
# notice that there is no column label and 
# the index is set automatically

<class 'pandas.core.frame.DataFrame'>


0         1         2
0  0.156609 -1.553114 -1.153612
1  0.939583 -0.684207 -0.471353
2 -1.704274 -1.916492  1.182103

In [6]:
# Let use a 2-D arrays?
ax = np.array(np.random.randn(9)).reshape(3,3)
print (ax)
df0 = pd.DataFrame(ax)
df0

[[ 0.78175536 -0.78182392  0.47415842]
 [ 0.2992212  -0.86199811  0.08339434]
 [-0.97546484  1.19893314  0.80690741]]


0         1         2
0  0.781755 -0.781824  0.474158
1  0.299221 -0.861998  0.083394
2 -0.975465  1.198933  0.806907

In [47]:
# And now, lets use a dictionary as input

dict1 = {'A':a1, 'B':a2}
df1 = pd.DataFrame(dict1) 
df1
# note headings

A         B
0  0.156609  0.939583
1 -1.553114 -0.684207
2 -1.153612 -0.471353

In [48]:
# Lets add another column
df1['C']=a3
df1

A         B         C
0  0.156609  0.939583 -1.704274
1 -1.553114 -0.684207 -1.916492
2 -1.153612 -0.471353  1.182103

In [49]:
# What is we assign a list instead of numpy array?
df1['L'] = ["List", "of", "words"]
print ("The column L is a ",type (df1['L']))
df1

The column L is a  <class 'pandas.core.series.Series'>


A         B         C      L
0  0.156609  0.939583 -1.704274   List
1 -1.553114 -0.684207 -1.916492     of
2 -1.153612 -0.471353  1.182103  words

In [50]:
# Introducing a pandas.Series
# Its like an np.array but it has its own index
s = pd.Series([1,np.nan,3])
s2 = pd.Series([2, 3, 4], index = ['a','b','c'])
print (s)
print()
print (s2)

0    1.0
1    NaN
2    3.0
dtype: float64

a    2
b    3
c    4
dtype: int64


In [51]:
# We will add the Series s to the table as column S
df1['S'] = s
df1

A         B         C      L    S
0  0.156609  0.939583 -1.704274   List  1.0
1 -1.553114 -0.684207 -1.916492     of  NaN
2 -1.153612 -0.471353  1.182103  words  3.0

In [52]:
#s2.index = df1.index

In [57]:
# This time, we will use Series s2, 
# which has a different index
df1['S2']=s2
print (df1)

          A         B         C      L    S  S2
0  0.156609  0.939583 -1.704274   List  1.0 NaN
1 -1.553114 -0.684207 -1.916492     of  NaN NaN
2 -1.153612 -0.471353  1.182103  words  3.0 NaN


In [66]:
# But if we create a new dataframe, we can add the data but with the new index
df2 = pd.DataFrame(s2)
df2['A']= a1
df2['B']=a2
df2['C']=a3
print (df2)

   0         A         B         C
a  2  0.156609  0.939583 -1.704274
b  3 -1.553114 -0.684207 -1.916492
c  4 -1.153612 -0.471353  1.182103


In [67]:
# You can extract rows by position or label
print (df2[1:3])
print()
print (df2['a':'b'])

   0         A         B         C
b  3 -1.553114 -0.684207 -1.916492
c  4 -1.153612 -0.471353  1.182103

   0         A         B         C
a  2  0.156609  0.939583 -1.704274
b  3 -1.553114 -0.684207 -1.916492


In [68]:
# recall
df1

A         B         C      L    S  S2
0  0.156609  0.939583 -1.704274   List  1.0 NaN
1 -1.553114 -0.684207 -1.916492     of  NaN NaN
2 -1.153612 -0.471353  1.182103  words  3.0 NaN

In [69]:
# Renaming a column
df1 = df1.rename(columns = {'L':'D'})
df1

A         B         C      D    S  S2
0  0.156609  0.939583 -1.704274   List  1.0 NaN
1 -1.553114 -0.684207 -1.916492     of  NaN NaN
2 -1.153612 -0.471353  1.182103  words  3.0 NaN

In [72]:
# And delete column
del df1['S2']
df1

A         B         C      D    S
0  0.156609  0.939583 -1.704274   List  1.0
1 -1.553114 -0.684207 -1.916492     of  NaN
2 -1.153612 -0.471353  1.182103  words  3.0

In [74]:
# Example: view only a column
df1['A']

0    0.156609
1   -1.553114
2   -1.153612
Name: A, dtype: float64

In [78]:
# A list of colums

print (type(df1['A']))
df1[['A','D']]
# Notice the data structure of the column is a Series, not an array


<class 'pandas.core.series.Series'>


A      D
0  0.156609   List
1 -1.553114     of
2 -1.153612  words

In the 10 min Pandas Guide, you will see many ways to view, slice a dataframe

* view/slice by rows, eg df[1:3], etc.

* view by index location, see df.iloc (iloc)

* view by ranges of labels, ie index label 2 to 5, or dates feb 3 to feb 25, see df.loc (loc)
 
* view a single row by the index df.xs (xs)

* filtering rows that have certain conditions
* add column
* add row

* How to change the index

and more...

In [82]:
print (df1[0:2])  # ok
# df1[1]  # not ok

#What will this do?
# print df1[0:2]
# print (df1[0:2][0:1])

          A         B         C     D    S
0  0.156609  0.939583 -1.704274  List  1.0
1 -1.553114 -0.684207 -1.916492    of  NaN


## Finance example

### Now, lets get some data in a CSV file is like this.

See https://www.quantshare.com/sa-43-10-ways-to-download-historical-stock-quotes-data-for-free


In [83]:
# We can use this 'pd.read_csv' method with the yahoo url, 1 year back with google's service
dfg = pd.read_csv('https://www.google.com/finance/historical?output=csv&q=goog')
dfa = pd.read_csv('https://www.google.com/finance/historical?output=csv&q=aapl')

In [84]:
# some viewing options
print ('aaple:')
print (dfa[0:3])
print ('google:')
print (dfg.head(10))
# dfg[0:10]
# dfg
print (dfg.tail(3))

aaple:
        Date    Open    High     Low   Close    Volume
0   5-Sep-17  163.75  164.25  160.56  162.08  29536314
1   1-Sep-17  164.80  164.94  163.63  164.05  16591051
2  31-Aug-17  163.64  164.52  163.48  164.00  26785096
google:
        Date    Open    High     Low   Close   Volume
0   5-Sep-17  933.08  937.00  921.96  928.45  1348292
1   1-Sep-17  941.13  942.48  935.15  937.34   947374
2  31-Aug-17  931.76  941.98  931.76  939.33  1582579
3  30-Aug-17  920.05  930.82  919.65  929.57  1301225
4  29-Aug-17  905.10  923.33  905.00  921.29  1185564
5  28-Aug-17  916.00  919.24  911.87  913.81  1086484
6  25-Aug-17  923.49  925.56  915.50  915.89  1053376
7  24-Aug-17  928.66  930.84  915.50  921.28  1270306
8  23-Aug-17  921.93  929.93  919.36  927.00  1090248
9  22-Aug-17  912.72  925.86  911.48  924.69  1166737
          Date    Open    High     Low   Close   Volume
247  12-Sep-16  755.13  770.29  754.00  769.02  1310986
248   9-Sep-16  770.10  773.24  759.66  759.66  1885496
249

In [85]:
# if you want the index by itelf, use .index
# dfg.index
dfg.index[2:20]

RangeIndex(start=2, stop=20, step=1)

In [86]:
# Here is a list of the columns, so you know the names of the columns
dfg.columns
#dfg.columns[2]
#dfg.columns[2:4]

#try this: (element of row 3 from every column)

#for i in dfg.columns:
#    print (i, df0_goog[i][3])


Index(['Date', 'Open', 'High', 'Low', 'Close', 'Volume'], dtype='object')

In [87]:
# If you want the values in an np array
npg = dfg.values
print (npg, type(npg))

[['5-Sep-17' 933.08 937.0 921.96 928.45 1348292]
 ['1-Sep-17' 941.13 942.48 935.15 937.34 947374]
 ['31-Aug-17' 931.76 941.98 931.76 939.33 1582579]
 ..., 
 ['12-Sep-16' 755.13 770.29 754.0 769.02 1310986]
 ['9-Sep-16' 770.1 773.24 759.66 759.66 1885496]
 ['8-Sep-16' 778.59 780.35 773.58 775.32 1270264]] <class 'numpy.ndarray'>


### Getting or Viewing Data within a DataFrame
Note: While standard Python / Numpy expressions for selecting and setting are intuitive and come in handy for interactive work, for production code, we recommend the optimized pandas data access methods, .at, .iat, .loc, .iloc and .ix. (from 10 min guide to Pandas)


In [89]:
# Lets print the 1 column, which show close prices of Google
# This is a new data frame (like a new table)
dfg['Close'][0:5]

0    928.45
1    937.34
2    939.33
3    929.57
4    921.29
Name: Close, dtype: float64

In [27]:
# Lets print the 2 column, and only 3 down, which show close prices of Google
# Instead of one column name, we have a python list ['Date',Close']
dfg[['Date','Close']][0:3]

Date   Close
0  21-Jun-17  959.45
1  20-Jun-17  950.63
2  19-Jun-17  957.37

In [28]:
# A slice: by rows (row numbers)
print (dfg[1:5])

        Date    Open    High     Low   Close   Volume
1  20-Jun-17  957.52  961.62  950.01  950.63  1125990
2  19-Jun-17  949.96  959.99  949.05  957.37  1533336
3  16-Jun-17  940.00  942.04  931.60  939.78  3094711
4  15-Jun-17  933.97  943.34  924.44  942.31  2133050


In [29]:
# Getting a cross section with .loc - BY VALUES of the index and headers
# Introduce loc: this will get you a cross section of the table by label ran
# df.loc[a:b, x:y], by rows and column location

print (dfg.loc[1:4,'Open':'Close'])

# recall:
# dfg[a:b] by rows
# dfg[[col]] or df[[list of col]] by columns
# df.loc[a:b, x:y], by rows and column location
# df.iloc[3:5,0:2], by slicing by specific position - BY POSITION in the table

print (dfg.iloc[1:4,4:6])


     Open    High     Low   Close
1  957.52  961.62  950.01  950.63
2  949.96  959.99  949.05  957.37
3  940.00  942.04  931.60  939.78
4  933.97  943.34  924.44  942.31
    Close   Volume
1  950.63  1125990
2  957.37  1533336
3  939.78  3094711


In [30]:
#n Data only from row with index value '3'
print (dfg.iloc[3])

Date      16-Jun-17
Open            940
High         942.04
Low           931.6
Close        939.78
Volume      3094711
Name: 3, dtype: object


In [31]:
# iloc will accept 'lists' of position numbers
dfg.iloc[[1,2,4],[0,2]]

Date    High
1  20-Jun-17  961.62
2  19-Jun-17  959.99
4  15-Jun-17  943.34

In [32]:
# iloc will accept a range with ':', just like numpy
dfg.iloc[1:3,:]

Date    Open    High     Low   Close   Volume
1  20-Jun-17  957.52  961.62  950.01  950.63  1125990
2  19-Jun-17  949.96  959.99  949.05  957.37  1533336

In [35]:
# Can also return specific value
print (dfg.iloc[2,1])
# same as above but faster for one single scaler value
print (dfg.iat[2,1])

949.96
949.96


### Basic Statistics

In [36]:
# A quick way to get statistics
dfg.describe()
# dfg.describe()['A'][1]
# dfg.describe()[2:3]

Open        High         Low       Close        Volume
count  251.000000  251.000000  251.000000  251.000000  2.510000e+02
mean   811.345139  816.271036  806.340159  811.818645  1.540603e+06
std     67.900802   68.045718   67.343274   67.890060  6.777015e+05
min    671.000000  672.300000  663.280000  668.260000  5.874210e+05
25%    772.450000  776.260000  767.150000  771.535000  1.108766e+06
50%    795.470000  801.190000  791.190000  795.370000  1.362115e+06
75%    831.635000  836.015000  827.870000  831.580000  1.704087e+06
max    984.500000  988.250000  977.200000  983.680000  4.745183e+06

In [99]:
print (dfg[0:5])
dfg.sort_index(axis=0, ascending=False).head() # by index or column, try axis = 0

        Date    Open    High     Low   Close   Volume
0   5-Sep-17  933.08  937.00  921.96  928.45  1348292
1   1-Sep-17  941.13  942.48  935.15  937.34   947374
2  31-Aug-17  931.76  941.98  931.76  939.33  1582579
3  30-Aug-17  920.05  930.82  919.65  929.57  1301225
4  29-Aug-17  905.10  923.33  905.00  921.29  1185564


Date    Open    High     Low   Close   Volume
249   8-Sep-16  778.59  780.35  773.58  775.32  1270264
248   9-Sep-16  770.10  773.24  759.66  759.66  1885496
247  12-Sep-16  755.13  770.29  754.00  769.02  1310986
246  13-Sep-16  764.48  766.22  755.80  759.69  1395046
245  14-Sep-16  759.61  767.68  759.11  762.49  1094490

In [100]:
# sort by value
dfg[0:5].sort_values(by='Open')

Date    Open    High     Low   Close   Volume
4  29-Aug-17  905.10  923.33  905.00  921.29  1185564
3  30-Aug-17  920.05  930.82  919.65  929.57  1301225
2  31-Aug-17  931.76  941.98  931.76  939.33  1582579
0   5-Sep-17  933.08  937.00  921.96  928.45  1348292
1   1-Sep-17  941.13  942.48  935.15  937.34   947374

In [101]:
# Transpose in Pandas
print (dfg.describe()[2:3])
print(dfg.describe()[2:3].T)

# try other methods like sum, mean, etc. in the same way.

          Open       High        Low      Close         Volume
std  70.959824  71.263562  70.339006  71.108088  668696.394467
                  std
Open        70.959824
High        71.263562
Low         70.339006
Close       71.108088
Volume  668696.394467


### Masks and Boolean Indexing

In [40]:
dfg[0:10]

Date    Open    High     Low   Close   Volume
0  21-Jun-17  953.64  960.10  950.76  959.45  1200693
1  20-Jun-17  957.52  961.62  950.01  950.63  1125990
2  19-Jun-17  949.96  959.99  949.05  957.37  1533336
3  16-Jun-17  940.00  942.04  931.60  939.78  3094711
4  15-Jun-17  933.97  943.34  924.44  942.31  2133050
5  14-Jun-17  959.92  961.15  942.25  950.76  1489715
6  13-Jun-17  951.91  959.98  944.09  953.40  2013337
7  12-Jun-17  939.56  949.36  915.23  942.90  3763529
8   9-Jun-17  984.50  984.50  935.63  949.83  3309389
9   8-Jun-17  982.35  984.57  977.20  983.41  1481916

In [41]:
# mask 1
mg1 = dfg['Open'][0:10]>941
print (mg1)
# dfg.Open[0:10]>941    # same thing

dfg[dfg['Open']>941][0:10]
# shows only rows with opening price greater than 941

0     True
1     True
2     True
3    False
4    False
5     True
6     True
7    False
8     True
9     True
Name: Open, dtype: bool


Date    Open    High     Low   Close   Volume
0   21-Jun-17  953.64  960.10  950.76  959.45  1200693
1   20-Jun-17  957.52  961.62  950.01  950.63  1125990
2   19-Jun-17  949.96  959.99  949.05  957.37  1533336
5   14-Jun-17  959.92  961.15  942.25  950.76  1489715
6   13-Jun-17  951.91  959.98  944.09  953.40  2013337
8    9-Jun-17  984.50  984.50  935.63  949.83  3309389
9    8-Jun-17  982.35  984.57  977.20  983.41  1481916
10   7-Jun-17  979.65  984.15  975.77  981.08  1453874
11   6-Jun-17  983.16  988.25  975.14  976.57  1814624
12   5-Jun-17  976.55  986.91  975.10  983.68  1252106

In [42]:
# mask 2
mg2 = dfg[0:10]>941
print (mg2)
print (dfg[dfg>941].head(10))
# replaces every value in the entire table with NaN if the value of below 941

   Date   Open  High    Low  Close  Volume
0  True   True  True   True   True    True
1  True   True  True   True   True    True
2  True   True  True   True   True    True
3  True  False  True  False  False    True
4  True  False  True  False   True    True
5  True   True  True   True   True    True
6  True   True  True   True   True    True
7  True  False  True  False   True    True
8  True   True  True  False   True    True
9  True   True  True   True   True    True
        Date    Open    High     Low   Close   Volume
0  21-Jun-17  953.64  960.10  950.76  959.45  1200693
1  20-Jun-17  957.52  961.62  950.01  950.63  1125990
2  19-Jun-17  949.96  959.99  949.05  957.37  1533336
3  16-Jun-17     NaN  942.04     NaN     NaN  3094711
4  15-Jun-17     NaN  943.34     NaN  942.31  2133050
5  14-Jun-17  959.92  961.15  942.25  950.76  1489715
6  13-Jun-17  951.91  959.98  944.09  953.40  2013337
7  12-Jun-17     NaN  949.36     NaN  942.90  3763529
8   9-Jun-17  984.50  984.50     NaN  949

In [43]:
# another way to filter is with isin()
# syntax only
# df2[df2['E'].isin(['two','four'])]

In [44]:
# like Numpy, sometimes you need an actual copy, not a view or slice of the same data
dfg2 = dfg.copy()
dfg2[0:5]

Date    Open    High     Low   Close   Volume
0  21-Jun-17  953.64  960.10  950.76  959.45  1200693
1  20-Jun-17  957.52  961.62  950.01  950.63  1125990
2  19-Jun-17  949.96  959.99  949.05  957.37  1533336
3  16-Jun-17  940.00  942.04  931.60  939.78  3094711
4  15-Jun-17  933.97  943.34  924.44  942.31  2133050

### Setting Values


In [49]:
# Recall
print(dfg.head(10))

        Date    Open    High     Low   Close   Volume
0  21-Jun-17  953.64  960.10  950.76  959.45  1200693
1  20-Jun-17  957.52  961.62  950.01  950.63  1125990
2  19-Jun-17  949.96  959.99  949.05  957.37  1533336
3  16-Jun-17  940.00  942.04  931.60  939.78  3094711
4  15-Jun-17  933.97  943.34  924.44  942.31  2133050
5  14-Jun-17  959.92  961.15  942.25  950.76  1489715
6  13-Jun-17  951.91  959.98  944.09  953.40  2013337
7  12-Jun-17  939.56  949.36  915.23  942.90  3763529
8   9-Jun-17  984.50  984.50  935.63  949.83  3309389
9   8-Jun-17  982.35  984.57  977.20  983.41  1481916


In [52]:
# All the ways to view (by location, by index, iat, etc) can also be used to set values
dfg['Volume'] = dfg['Volume']/1000.0
print(dfg.head(10))

        Date    Open    High     Low   Close    Volume
0  21-Jun-17  953.64  960.10  950.76  959.45  1.200693
1  20-Jun-17  957.52  961.62  950.01  950.63  1.125990
2  19-Jun-17  949.96  959.99  949.05  957.37  1.533336
3  16-Jun-17  940.00  942.04  931.60  939.78  3.094711
4  15-Jun-17  933.97  943.34  924.44  942.31  2.133050
5  14-Jun-17  959.92  961.15  942.25  950.76  1.489715
6  13-Jun-17  951.91  959.98  944.09  953.40  2.013337
7  12-Jun-17  939.56  949.36  915.23  942.90  3.763529
8   9-Jun-17  984.50  984.50  935.63  949.83  3.309389
9   8-Jun-17  982.35  984.57  977.20  983.41  1.481916


In [53]:
dfg['Volume'] = 0
print(dfg.head(10))

        Date    Open    High     Low   Close  Volume
0  21-Jun-17  953.64  960.10  950.76  959.45       0
1  20-Jun-17  957.52  961.62  950.01  950.63       0
2  19-Jun-17  949.96  959.99  949.05  957.37       0
3  16-Jun-17  940.00  942.04  931.60  939.78       0
4  15-Jun-17  933.97  943.34  924.44  942.31       0
5  14-Jun-17  959.92  961.15  942.25  950.76       0
6  13-Jun-17  951.91  959.98  944.09  953.40       0
7  12-Jun-17  939.56  949.36  915.23  942.90       0
8   9-Jun-17  984.50  984.50  935.63  949.83       0
9   8-Jun-17  982.35  984.57  977.20  983.41       0


In [57]:
dfg.iat[0,1] = 0
dfg.head(3)

Date    Open    High     Low   Close  Volume
0  21-Jun-17    0.00  960.10  950.76  959.45       0
1  20-Jun-17  957.52  961.62  950.01  950.63       0
2  19-Jun-17  949.96  959.99  949.05  957.37       0

In [ ]:
# Comments on dropping and filling NaN values
# A view where we drop any rows with value NnN
# dfg.dropna(how='any')  # this would be used to drop rows with Nan
# df1.fillna(value=5)    # this would be used to fill NaN values with 5

### More Statistics and Operations

In [60]:
dfg.mean()   # mean by column, also try var()

Open      807.545777
High      816.271036
Low       806.340159
Close     811.818645
Volume      0.000000
dtype: float64

In [62]:
dfg[0:5].mean(1)
# dfg.mean(axis = 1)

0    574.062
1    763.956
2    763.274
3    750.684
4    748.812
dtype: float64

In [65]:
# Use the apply method to perform calculations on every element
dfg2[0:10].apply(np.cumsum)

Date     Open     High  \
0                                          21-Jun-17   953.64   960.10   
1                                 21-Jun-1720-Jun-17  1911.16  1921.72   
2                        21-Jun-1720-Jun-1719-Jun-17  2861.12  2881.71   
3               21-Jun-1720-Jun-1719-Jun-1716-Jun-17  3801.12  3823.75   
4      21-Jun-1720-Jun-1719-Jun-1716-Jun-1715-Jun-17  4735.09  4767.09   
5  21-Jun-1720-Jun-1719-Jun-1716-Jun-1715-Jun-171...  5695.01  5728.24   
6  21-Jun-1720-Jun-1719-Jun-1716-Jun-1715-Jun-171...  6646.92  6688.22   
7  21-Jun-1720-Jun-1719-Jun-1716-Jun-1715-Jun-171...  7586.48  7637.58   
8  21-Jun-1720-Jun-1719-Jun-1716-Jun-1715-Jun-171...  8570.98  8622.08   
9  21-Jun-1720-Jun-1719-Jun-1716-Jun-1715-Jun-171...  9553.33  9606.65   

       Low    Close    Volume  
0   950.76   959.45   1200693  
1  1900.77  1910.08   2326683  
2  2849.82  2867.45   3860019  
3  3781.42  3807.23   6954730  
4  4705.86  4749.54   9087780  
5  5648.11  5700.30  10577495  
6  6592.20  6653.70  12590832  
7  7507.43  7596.60  16354361  
8  8443.06  8546.43  19663750  
9  9420.26  9529.84  21145666

In [104]:
# Reload
dfg = pd.read_csv('https://www.google.com/finance/historical?output=csv&q=goog')
dfa = pd.read_csv('https://www.google.com/finance/historical?output=csv&q=aapl')
dfm = pd.read_csv('https://www.google.com/finance/historical?output=csv&q=msft')
#dfd = pd.read_csv('https://www.google.com/finance/historical?output=csv&q=dis')
dfn = pd.read_csv('https://www.google.com/finance/historical?output=csv&q=nke')
dfb = pd.read_csv('https://www.google.com/finance/historical?output=csv&q=ba')

In [105]:
print (dfb.head())

        Date    Open    High     Low   Close   Volume
0   5-Sep-17  240.30  240.36  236.23  237.00  3869251
1   1-Sep-17  239.66  242.53  239.17  240.33  3271270
2  31-Aug-17  241.00  241.10  238.41  239.66  3075798
3  30-Aug-17  240.99  241.44  239.00  240.46  2801097
4  29-Aug-17  237.21  241.17  236.08  240.49  3535465


In [106]:
# Rename columns
dfg = dfg.rename(columns = {'Close':'GOOG'})
#print (dfg.head())

dfa = dfa.rename(columns = {'Close':'AAPL'})
#print (dfa.head())

dfm = dfm.rename(columns = {'Close':'MSFT'})
#print (dfm.head())

#dfd = dfd.rename(columns = {'Close':'DIS'})
#print (dfd.head())

dfn = dfn.rename(columns = {'Close':'NKE'})
#print (dfn.head())

dfb = dfb.rename(columns = {'Close':'BA'})
print (dfb.head())


        Date    Open    High     Low      BA   Volume
0   5-Sep-17  240.30  240.36  236.23  237.00  3869251
1   1-Sep-17  239.66  242.53  239.17  240.33  3271270
2  31-Aug-17  241.00  241.10  238.41  239.66  3075798
3  30-Aug-17  240.99  241.44  239.00  240.46  2801097
4  29-Aug-17  237.21  241.17  236.08  240.49  3535465


In [107]:
# Lets merge some tables
df = dfg[['Date','GOOG']].merge(dfa[['Date','AAPL']])
df = df.merge(dfm[['Date','MSFT']])
#df = df.merge(dfd[['Date','DIS']])
df = df.merge(dfn[['Date','NKE']])
df = df.merge(dfb[['Date','BA']])
print (df[0:10])



        Date    GOOG    AAPL   MSFT    NKE      BA
0   5-Sep-17  928.45  162.08  73.61  53.01  237.00
1   1-Sep-17  937.34  164.05  73.94  53.36  240.33
2  31-Aug-17  939.33  164.00  74.77  52.81  239.66
3  30-Aug-17  929.57  163.35  74.01  52.56  240.46
4  29-Aug-17  921.29  162.91  73.05  52.73  240.49
5  28-Aug-17  913.81  161.47  72.83  53.73  237.18
6  25-Aug-17  915.89  159.86  72.82  53.90  235.89
7  24-Aug-17  921.28  159.27  72.69  53.84  238.74
8  23-Aug-17  927.00  159.98  72.72  53.61  238.09
9  22-Aug-17  924.69  159.78  73.16  54.13  239.75


In [110]:
# show a correlation matrix (pearson)
crl = df.corr()
print (crl)
print()
print (crl.sort_values(by='GOOG',ascending=False))

df[30:120].corr()


          GOOG      AAPL      MSFT       NKE        BA
GOOG  1.000000  0.898836  0.925628  0.416602  0.820643
AAPL  0.898836  1.000000  0.923779  0.556489  0.893996
MSFT  0.925628  0.923779  1.000000  0.491339  0.946342
NKE   0.416602  0.556489  0.491339  1.000000  0.570211
BA    0.820643  0.893996  0.946342  0.570211  1.000000

          GOOG      AAPL      MSFT       NKE        BA
GOOG  1.000000  0.898836  0.925628  0.416602  0.820643
MSFT  0.925628  0.923779  1.000000  0.491339  0.946342
AAPL  0.898836  1.000000  0.923779  0.556489  0.893996
BA    0.820643  0.893996  0.946342  0.570211  1.000000
NKE   0.416602  0.556489  0.491339  1.000000  0.570211


GOOG      AAPL      MSFT       NKE        BA
GOOG  1.000000  0.756714  0.915886 -0.322761  0.679831
AAPL  0.756714  1.000000  0.625128 -0.410547  0.209578
MSFT  0.915886  0.625128  1.000000 -0.061586  0.822941
NKE  -0.322761 -0.410547 -0.061586  1.000000  0.268532
BA    0.679831  0.209578  0.822941  0.268532  1.000000

## Homework Section

In [187]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

#### 1) Read in a CSV file called 'data3.csv' into a dataframe called df.
#### 2) Display the first few lines

#### 3)  If you inspect the whole file in an editor, you will see that the bottom 8 lines in the frame need to be dropped.  Remove those 8 lines from df.  You can also use df.tail() to see the last lines.

#### 4) Drop columns: drop the columns Area Id, Variable Id, Symbol, and Md.  Hint Google 'delete a column in pandas 


#### 5) # Convert the year column to pandas datetime?

Convert the 'Year' column string values to pandas datetime objects, where only the year is specified. To do this, simply run the code below (please review it carefully and make sure you understand what happens):

df['Year'] = pd.to_datetime(pd.Series(df['Year']).astype(int),format='%Y').dt.year
Run df.tail() to see if you get what you expect

#### 6) Print rows where Area is Iceland

#### 7) Print the years when the National Rainfall Index (NRI) was greater than 950 or less than 900 in Iceland, by using this code.  Put it ina dataframe called df_temp so you don't change the original 'df'

#### 8) Get al the rows of the original df area is United States of America 

a) put the USA into a df_usa data frame

b) Add the columns to ['GDP','NRI','PD','Area','Population']

c) delending on the 'Variable Name' put the value into respective column 'GDP','NRI','PD','Area', or 'Population'

Look into this syntax in case its helpful:
df['color'] = np.where(df['Set']=='Z', 'green', 'red')

d) Find the maximum value and minimum value of the 'NRI' column in the US (using pandas methods). What years do the min and max values occur?

#### 9) Show general statistics for columns 'GDP','NRI','PD', and 'Population'

#### 10 a) Show a 3 x 3 correlation matrix for Nike, Apple, and Disney stock prices over the past year

#### 10b) Show the same correlation matrix but over different time periods, 

i) the last 20 days
ii) the last 80 days

#### 11) Change the code so that it accepts a list of any stock symbols, ie ['NKE', 'APPL', 'DIS', ... ] and creates a correlation matrix for the time period of the past 100 days
